runtime/lua/man.lua

```diff
@@ -0,0 +1,141 @@
+local function highlight_formatted(line, linenr)
+  local chars = {}
+  local prev_char = ''
+  local overstrike, escape = false, false
+  local hls = {} -- Store highlight groups as { attr, start, end }
+  local NONE, BOLD, UNDERLINE, ITALIC = 0, 1, 2, 3
+  local hl_groups = {[BOLD]="manBold", [UNDERLINE]="manUnderline", [ITALIC]="manItalic"}
+  local attr = NONE
+  local byte = 0 -- byte offset
+
+  local function end_attr_hl(attr)
+    for i, hl in ipairs(hls) do
+      if hl[1] == attr and hl[3] == -1 then
+        hl[3] = byte
+        hls[i] = hl
+      end
+    end
+  end
+
+  local function add_attr_hl(code)
+    local on = true
+    if code == 0 then
+      attr = NONE
+      on = false
+    elseif code == 1 then
+      attr = BOLD
+    elseif code == 21 or code == 22 then
+      attr = BOLD
+      on = false
+    elseif code == 3 then
+      attr = ITALIC
+    elseif code == 23 then
+      attr = ITALIC
+      on = false
+    elseif code == 4 then
+      attr = UNDERLINE
+    elseif code == 24 then
+      attr = UNDERLINE
+      on = false
+    else
+      attr = NONE
+      return
+    end
+
+    if on then
+      hls[#hls + 1] = {attr, byte, -1}
+    else
+      if attr == NONE then
+        for a, _ in pairs(hl_groups) do
+          end_attr_hl(a)
+        end
+      else
+        end_attr_hl(attr)
+      end
+    end
+  end
+
+  -- Break input into UTF8 characters
+  for char in line:gmatch("[^\128-\191][\128-\191]*") do
+    if overstrike then
+      local last_hl = hls[#hls]
+      if char == prev_char then
+        if char == '_' and attr == UNDERLINE and last_hl and last_hl[3] == byte then
+          -- This underscore is in the middle of an underlined word
```

Comment:
When would this occur? Could you provide some sample input to make the comment more clear?

And also, is there any precedence for this behaviour on when to underline vs bold in regards to an overstriken underscore? To me, it seems intuitive that it would always become bolded.

Reply:
This doesn't occur too often, but one example would be the man page for `mawk`. View it with `less`, search for `opt_expr`, and you'll see an underlined word with an underscore. The behavior of checking the previous overstrike and defaulting to bold is essentially what `less` does too.

Reply:
Very odd, I'd expect that to be a space, followed by a backspace and then an underline, not a overstriken underline. Otherwise LGTM but do we need the `last_hl and last_hl[3] == byte` check?